### PR TITLE
[New Model]: Planned Production Output

### LINKED ISSUE #290: [New Model]: Planned Production Output
## Model Description
PURIS (UC) wants to provide further information to evaluate the supply situation. One of this information is the planned production (quantity and date time matchings). It is complementary to product stock (find it [here](https://github.com/eclipse-tractusx/sldt-semantic-models/tree/main/io.catenax.product_stock/1.0.0)). Together they should support identifying and mitigating shortages more actively.

## Relevant Standards
None

## Example Data
``` json
{
  "positions": [
    {
      "lastUpdatedOnDateTime": "2023-04-01T14:23:00+01:00",
      "allocatedProductionOutputs": [
        {
          "plannedProductionQuantity": {
            "unitOfMeasure": "GRM",
            "quantityNumber": 2.5
          },
          "productionSiteBpns": "BPNS0123456789ZZ",
          "estimatedTimeOfCompletion": "2023-04-01T14:23:00+01:00"
        }
      ],
      "orderPositionReference": {
        "supplierOrderId": "M-Nbr-4711",
        "customerOrderId": "C-Nbr-4711",
        "customerOrderPositionId": "PositionId-01"
      }
    }
  ],
  "materialNumberCatenaX": "urn:uuid:48878d48-6f1d-47f5-8ded-a441d0d879df",
  "materialNumberCustomer": "MNR-7307-AU340474.002",
  "materialNumberSupplier": "MNR-8101-ID146955.001"
}
```

## MS1 Criteria

- [x] The proposed aspect model does not exist already
- [x] The proposed aspect model does not extend an existing aspect but introduces completely new functionality
- [x] The proposal references a Catena-X use case
- [x] Relevant standards are mentioned/linked
- [x] A first draft of the data structure is provided in an example

### PR DESCRIPTION
**Pull request by tom-rm-meyer-ISST**
## Description
Planned Production Output for ad hoc information and monitoring in supply chain relation.

Closes #290 

## MS2 Criteria
(to be filled out by PR reviewer)
- [x] the model **validates** with the BAMM SDS SDK in the version specified in the Readme.md of this repository by the time of the MS2 check  (e.g., 'java -jar bamm-cli.jar -i \<path-to-aspect-model\> -v ). The  SAMM CLI is available [here](https://eclipse-esmf.github.io/esmf-developer-guide/tooling-guide/samm-cli.html) and in [GitHub](https://github.com/eclipse-esmf/esmf-sdk/releases/tag/v2.3.1)
- [x] use **Camel-Case** (e.g., "MyModelElement" or "TimeDifferenceGmtId", when in doubt follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)
- [x] the identifiers for all model elements **start with a capital letter** except for properties
- [x] the identifier for **properties starts with a small letter**
- [x] all model elements **at least contain the fields "preferred name" and "description"** in English language. The description must be comprehensible. It is not required to write full sentences but style should be consistent over the whole model
- [x] Property and the referenced Characteristic should not have the same name
- [x] the versioning in the URN **follows semantic versioning**, where minor version bumps are backwards compatible and major version bumps are not backwards compatible. 
- [x] use **abbreviations only when necessary** and if these are sufficiently common
- [x] **avoid redundant prefixes in property names** (consider adding properties to an enclosing Entity or even adapt the namespace of the model elements, e.g., instead of having two properties `DismantlerId` and `DismantlerName` use an Entity `Dismantler` with the properties `name` and `id` or use a URN like `io.catenax.dismantler:0.0.1`)
- [x] fields `preferredName` and `description` are not the same
- [x] **`preferredName` should be human readable** and follow normal orthography (e.g., no camel case but normal word separation)
- [x] name of aspect is singular except if it only has one property which is a Collection, List or Set. In theses cases, the aspect name is plural.
- [x] units are referenced from the BAMM unit catalog whenever possible
- [x] **use constraints** to make known constraints from the use case explicit in the aspect model 
- [x] when relying on **external standards**, they are referenced through a **"see"** element
- [x] all properties with an [simple type](https://eclipse-esmf.github.io/samm-specification/2.0.0/datatypes.html) have an example value
- [x] metadata.json exists with status "release"
- [x] generated json schema validates against example json payload
- [x] file RELEASE_NOTES.md exists and contains entries for proposed model changes 
- [x] all contributors to this model are mentioned in copyright header of model file

## MS3 Criteria
(to be filled out by semantic modeling team before merge to main-branch)
- [x] All required reviewers have approved this PR (see reviewers section)
- [x] The new aspect (version) will be implemented by at least one data provider
- [x] The new aspect (version) will be consumed by at least one data consumer
- [x] There exists valid test data
- [x] In case of a new (incompatible) major version to an existing version, a migration strategy has been developed
- [x] The model has at least version '1.0.0'
- [x] The release date in the Release Note is set to the date of the MS3 approval

### CODE REVIEW

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+@prefix samm-e: <urn:samm:org.eclipse.esmf.samm:entity:2.0.0#>.
+@prefix unit: <urn:samm:org.eclipse.esmf.samm:unit:2.0.0#>.
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
+@prefix : <urn:samm:io.catenax.planned_production_output:1.0.0#>.
+
+:PlannedProductionOutput a samm:Aspect;
+    samm:preferredName "Planned Production Output of a Supplier"@en;
+    samm:description "This aspect represents the remaining and planned production output of a supplier. A planned production output is described by a quantity, the site of the supplier and the date with time at which a production of a certain material for a certain customer is planned to be finished. TODO: check if positions need to be considered somehow?"@en;
```
Comment (bs-jokri) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:12]: Todo will be removed?

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  samm:optional "true"^^xsd:boolean
+] [
+  samm:property :materialNumberSupplier;
+  samm:optional "true"^^xsd:boolean
+]);
+    samm:operations ();
+    samm:events ().
+:positions a samm:Property;
+    samm:preferredName "Positions"@en;
+    samm:description "The positions refer to positions within an order of the given product (see material number supplier, materialNumber customer and material number Catena-x) for which the production outputs are transferred. The positions can be differentiated by the order position reference. If no order position reference is given, then there is only one position containing all planned production outputs of the product."@en;
```
Comment (bs-jokri) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:24]: Catena-X

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    samm:properties ([
+  samm:property :orderPositionReference;
+  samm:optional "true"^^xsd:boolean
+] :lastUpdatedOnDateTime :allocatedPlannedProductionOutputs).
+:Uuidv4Characteristic a samm:Characteristic;
+    samm:preferredName "UUID v4"@en;
+    samm:description "A version 4 UUID is a universally unique identifier that is generated using random 32 hexadecimal characters."@en;
+    samm:dataType xsd:string;
+    samm:see <https://tools.ietf.org/html/rfc4122>.
+:Uuidv4RegularExpression a samm-c:RegularExpressionConstraint;
```
Comment (bs-jokri) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:66]: this is now probably the 10th or so model with a uuid characteristic definition. Maybe we should extract this to a separate shared aspect and reused in the other models?

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "Unit of measurement for demand quantities."@en;
+    samm:characteristic :UnitOfMeasure;
+    samm:exampleValue "KGM".
+:UnitOfMeasure a samm-c:Enumeration;
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "The Enumeration lists a selection of allowed units of measure referenced by their official code according to Codes for Units of Measure Used in International Trade"@en;
+    samm:dataType xsd:string;
+    samm:see <https://tfig.unece.org/contents/recommendation-20.htm>;
+    samm-c:values ("GRM" "KGM" "TNE" "STN" "ONZ" "LBR" "CMT" "MTR" "KTM" "INH" "FOT" "YRD" "CMK" "MTK" "INK" "FTK" "YDK" "CMQ" "MTQ" "INQ" "FTQ" "YDQ" "MLT" "LTR" "HLT" "H87" "SET" "PR" "ZP" "KWH" "").
```
Comment (bs-jokri) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:164]: the empty entry is by intention? If so explain that in the description

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+:BPNSTrait a samm-c:Trait;
+    samm:preferredName "BPNS Trait"@en;
+    samm:description "Trait to ensure data format for BPNS."@en;
+    samm-c:baseCharacteristic :BPNS;
+    samm-c:constraint :BPNSRegularExpression.
+:BPNS a samm:Characteristic;
+    samm:preferredName "Business Partner Number Site"@en;
+    samm:description "BPNS represents a site which can be a production plant inside one location."@en;
+    samm:dataType xsd:string.
+:BPNSRegularExpression a samm-c:RegularExpressionConstraint;
```
Comment (bs-jokri) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:183]: same here as with the uuid. We could think about extracting that to a shared aspect as it is already defined in a lot of different models

**Review by bs-jokri — DISMISSED** [review]

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl (hunk 1 above)
Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:12]: Added description regarding positions.

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl (hunk 4 above)
Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:164]: Was not by intention.

Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl]: Opened PR #299  to reuse model without empty enum value

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl (hunk 3 above)
Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl]: replaced by MS3 shared model. I'll rebase the branch after merging the model.

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl (hunk 4 above)
Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl]: replaced by MS3 model. I'll rebase the branch after merging the model.

**Comment by bs-jokri** [comment]
@tom-rm-meyer-ISST can you rebase you branch on latest main so that we have the shared models and the branch validates? Also, what are the next steps. Can the model be reviewed?

**Comment by tom-rm-meyer-ISST** [comment]
> @tom-rm-meyer-ISST can you rebase you branch on latest main so that we have the shared models and the branch validates? Also, what are the next steps. Can the model be reviewed?

We want to incorporate the unit of measure first (see [PR](https://github.com/eclipse-tractusx/sldt-semantic-models/pull/299)), but we're somewhat in discussion. I would love to get both things done in this week.

As soon as the mentioned PR is fixed, I'll incorporate the shared unit of measure and write a mail to you regarding MS2 (and MS3).

File: io.catenax.planned_production_output/RELEASE_NOTES.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+All notable changes to this model will be documented in this file.
+
+## [Unreleased]
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/RELEASE_NOTES.md]: set to 2023-10-09

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    samm:description "Planned Production Outputs already allocated to the customer."@en;
+    samm:characteristic :AllocatedPlannedProductionOutputCollection.
+:OrderPositionReferenceCharacteristic a samm:Characteristic;
+    samm:preferredName "Order Position Reference Characteristic"@en;
+    samm:description "The order position reference includes reference numbers for the order and the position. As these information are meant for the customer, the customer reference numbers are mandatory."@en;
+    samm:dataType :OrderPositionReference.
+:DateTimeStamp a samm:Characteristic;
+    samm:preferredName "Date Time Stamp Characteristic"@en;
+    samm:description "The Date with the timezone."@en;
+    samm:dataType xsd:date;
```
Comment (agg3fe) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:91]: Need to change to datetime

File: io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    samm:description "This is an identifier for the position of an order defined by the customer. It's not a technical identifier, such as an UUID. This identifier is used within the business context when talking about the position."@en;
+    samm:characteristic :OrderPositionIdCharacteristic;
+    samm:exampleValue "PositionId-01".
+:plannedProductionQuantity a samm:Property;
+    samm:preferredName "Planned Production Quantity"@en;
+    samm:description "Quantity of a product that a supplier plans to produce on a certain date and time for a specific customer (that means that it will be allocated to a specific customer's \"allocated item stock\").\nQuantities that have already been produced are put on stock or are considered for delivery. Therefore those quantities MUST NOT be considered in the allocated production output to build a consistent view on the supply situation. "@en;
+    samm:characteristic :QuantityCharacteristic.
+:productionSiteBpns a samm:Property;
+    samm:preferredName "BPNS of Supplier's Production Site"@en;
+    samm:description "One planned and alllocated production output has always one site. The site is identified by the BPNS."@en;
```
Comment (agg3fe) [inline io.catenax.planned_production_output/1.0.0/PlannedProductionOutput.ttl:129]: typo `alllocated`

**Comment by tom-rm-meyer-ISST** [comment]
@agg3fe thanks for the review. Can we remove the old review from @bs-jokri ? Changes should have been incorporated.

**Comment by tom-rm-meyer-ISST** [comment]
> @agg3fe thanks for the review. Can we remove the old review from @bs-jokri ? Changes should have been incorporated.

BTW: incorporated your comments :)

File: io.catenax.planned_production_output/RELEASE_NOTES.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+All notable changes to this model will be documented in this file.
+
+## [1.0.0] - 2023-11-2
```
Comment (agg3fe) [inline io.catenax.planned_production_output/RELEASE_NOTES.md:4]: you can change the date to 2023-11-20

Comment (tom-rm-meyer-ISST) [reply io.catenax.planned_production_output/RELEASE_NOTES.md]: Done

**Review by agg3fe — APPROVED** [review]
LGTM

**Pull request merged**